File: school/summer2013/openmp/fortran/14_mat_mul_dynamic.f90
program mat_mul_dynamic
    integer i,j,k, a(0:N-1,0:N-1), b(0:N-1,0:N-1), c(0:N-1,0:N-1)

    c = 0
    do j=0, N-1
        do i=0,j
            a(i,j) = i+j+1
            b(i,j) = i+j+2
        end do
        do i=j+1,N-1
            a(i,j) = 0
            b(i,j) = 0
        end do
    end do

!$omp parallel do private(i,k) &
!$omp& schedule(dynamic,5)
    do j=0, N-1
        do k=0, j
            do i=k, j
                c(i,j) = c(i,j) + a(i,k)*b(k,j)
            end do
        end do
    end do
!$omp end parallel do

    print *, 'data = ', c(N-1,N-1)
end

File: summer2013/openmp/c/14_mat_mul_dynamic.c
#include <stdio.h>

int main()
{
	int i, j, k;
	int a[N][N], b[N][N], c[N][N];

	for(i=0; i<N; i++) {
		for(j=0; j<i; j++) {
			a[i][j] = b[i][j] = c[i][j] = 0;
		}
		for(j=i; j<N; j++) {
			a[i][j] = i+j+1;
			b[i][j] = i+j+2;
			c[i][j] = 0;
		}
	}

#pragma omp parallel for private(j,k) schedule(dynamic,5)
	for(i=0; i<N; i++)
		for(j=i; j<N; j++)
			for(k=i; k<j+1; k++)
				c[i][j] += a[i][k] * b[k][j];

	printf("data = %d\n", c[N-1][N-1]);
}
